ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -83,6 +93,14 @@ import qualified Ouroboros.Consensus.Util.SlotBounded as SB
   Byron blocks and headers
 -------------------------------------------------------------------------------}
 
+-- | Unfortunate Byron requires some information that is morally static but
+-- actually comes from configuration files on startup. We use reflection to
+-- provide this to type class instances.
+type ByronGiven = (
+    Given Crypto.ProtocolMagicId
+  , Given CC.Slot.EpochSlots
+  )
```

Comment:
It was done this way in `cardano-sl`, as an improvement over the original design that used template haskell. Both are less than ideal: with TH, each build is fixed to one protocol magic; with reflections, it can be changed by runtime configuration, but it's not at all clear whether one process is capable of using two different protocol magics.

Reply:
Would be considered to be old fashioned to suggest that we just use parameter passing for this?

Reply:
That'd be my ideal solution. But since the design uses the typeclass `HasHeader`, something that makes these parameters _seem_ static is needed (otherwise the instances couldn't be implemented).

Reply:
Something more like an ML module would be nice here. Then the particular values of protocol magic and epoch slots could determine a new type for a byron block.